### PR TITLE
Adding Nine Patch center mode

### PR DESCRIPTION
**Pull request by Lisapple**
This keep the texture centered (but clips if patch smaller than texture). Useful for control handles, tooltips with arrows, etc.

tooltip

Only OpenGL ES 3 since ES2 only supports default stretch mode.

### CODE REVIEW

File: drivers/gles3/shaders/canvas.glsl [hunk 1]
```diff
@@ -387,26 +387,37 @@ float map_ninepatch_axis(float pixel, float draw_size, float tex_pixel_size, flo
...
-			float src_area = draw_size - margin_begin - margin_end;
-			float dst_area = tex_size - margin_begin - margin_end;
+			
 			float scale = max(1.0, floor(src_area / max(dst_area, 0.0000001) + 0.5));
 
 			//convert to ratio
 			float ratio = (pixel - margin_begin) / src_area;
 			ratio = mod(ratio * scale, 1.0);
 			return (margin_begin + ratio * dst_area) * tex_pixel_size;
+		} else if (np_repeat==3) { //center
```
Comment (YeldhamDev) [inline drivers/gles3/shaders/canvas.glsl]: Add a spacing before and after the "==".

File: drivers/gles3/shaders/canvas.glsl [hunk 2]
```diff
@@ -387,26 +387,37 @@ float map_ninepatch_axis(float pixel, float draw_size, float tex_pixel_size, flo
...
+			float center = (src_area - dst_area) / 2.0;
+			if (pixel <= margin_begin+center) {
+				// stretch begin
+				return (margin_begin + 1) * tex_pixel_size;
+			} else if (pixel >= draw_size-margin_end-center) {
```
Comment (YeldhamDev) [inline drivers/gles3/shaders/canvas.glsl]: Spacing in the "-" as well.

File: drivers/gles3/shaders/canvas.glsl [hunk 3]
```diff
@@ -387,26 +387,37 @@ float map_ninepatch_axis(float pixel, float draw_size, float tex_pixel_size, flo
...
+			
 			float scale = max(1.0, floor(src_area / max(dst_area, 0.0000001) + 0.5));
 
 			//convert to ratio
 			float ratio = (pixel - margin_begin) / src_area;
 			ratio = mod(ratio * scale, 1.0);
 			return (margin_begin + ratio * dst_area) * tex_pixel_size;
+		} else if (np_repeat==3) { //center
+			float center = (src_area - dst_area) / 2.0;
+			if (pixel <= margin_begin+center) {
```
Comment (YeldhamDev) [inline drivers/gles3/shaders/canvas.glsl]: Spacing in this "+".

File: drivers/gles3/shaders/canvas.glsl (hunk 1 above)
Comment (Lisapple) [reply drivers/gles3/shaders/canvas.glsl]: Yep, sorry about that

**Comment by bojidar-bg** [comment]
Would be nice to also have this option on StyleRectTexture, as it also uses nine-patching.

File: drivers/gles3/shaders/canvas.glsl [hunk 4]
```diff
@@ -387,26 +387,37 @@ float map_ninepatch_axis(float pixel, float draw_size, float tex_pixel_size, flo
...
 			//convert to ratio
-			float ofs = mod((pixel - margin_begin), tex_size - margin_begin - margin_end);
+			float ofs = mod((pixel - margin_begin), dst_area);
 			//scale to source texture
 			return (margin_begin + ofs) * tex_pixel_size;
 		} else if (np_repeat == 2) { //tile fit
 			//convert to ratio
-			float src_area = draw_size - margin_begin - margin_end;
-			float dst_area = tex_size - margin_begin - margin_end;
+
```
Comment (YeldhamDev) [inline drivers/gles3/shaders/canvas.glsl]: Remove this extra line.

File: drivers/gles3/shaders/canvas.glsl [hunk 5]
```diff
@@ -387,26 +387,37 @@ float map_ninepatch_axis(float pixel, float draw_size, float tex_pixel_size, flo
...
 			//convert to ratio
 			float ratio = (pixel - margin_begin) / src_area;
 			ratio = mod(ratio * scale, 1.0);
 			return (margin_begin + ratio * dst_area) * tex_pixel_size;
+		} else if (np_repeat == 3) { //center
+			float center = (src_area - dst_area) / 2.0;
+			if (pixel <= margin_begin + center) {
+				// stretch begin
+				return (margin_begin + 1) * tex_pixel_size;
+			} else if (pixel >= draw_size - margin_end-center) {
```
Comment (YeldhamDev) [inline drivers/gles3/shaders/canvas.glsl]: Spaces around the other "-" too.

**Comment by YeldhamDev** [comment]
Remember to squash these commits: https://docs.godotengine.org/en/3.1/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase

**Review by mhilbrunner — APPROVED** [review]

**Comment by akien-mga** [comment]
For the reference, this hasn't been merged yet because it's specific to GLES3, and it would make rebasing the `vulkan` branch on `master` more difficult (or this feature would simply not work on vulkan as is).

It's now too late for 3.2 which is in release freeze, but if there's a lot of interest in this feature we could see to get it implemented for 4.0 with Vulkan and GLES2, and possibly backport a GLES2 + GLES3 implementation to the 3.2 stable branch, if it doesn't break compatibility.

**Comment by aaronfranke** [comment]
@Lisapple Is this still desired? If so, it needs to be rebased on the latest master branch, and might need to be redesigned to work with Vulkan.

Otherwise, abandoned pull requests will be closed in the future as announced [here](https://godotengine.org/article/headsup-vulkan-merged-master-branch-unstable).

**Comment by Lisapple** [comment]
I think it could be an useful addition, but since the current OpenGL ES 2.0 implementation does not support any NinePatch mode (neither Vulkan yet), it demands too much effort for this to work on Godot 4.0 for now.

**Comment by Lisapple** [comment]
I think it could be an useful addition, but since the current OpenGL ES 2.0 implementation does not support any NinePatch mode (neither Vulkan yet), it demands too much effort for this to work on Godot 4.0 for now.

**Pull request closed without merging**